compiler/src/dotty/tools/dotc/plugins/Plugin.scala

```diff
@@ -0,0 +1,176 @@
+package dotty.tools.dotc
+package plugins
+
+import core._
+import Contexts._
+import Phases._
+import dotty.tools.io._
+import transform.MegaPhase.MiniPhase
+
+import java.io.InputStream
+
+import scala.collection.mutable
+import scala.util.{ Try, Success, Failure }
+
+trait PluginPhase extends MiniPhase {
+  def runsBefore: Set[Class[_ <: Phase]] = Set.empty
+}
+
+trait Plugin {
+  /** The name of this plugin */
+  def name: String
+
+  /** A one-line description of the plugin */
+  def description: String
+
+  /** Is this plugin a research plugin?
+   *
+   *  Research plugin receives a phase plan and return a new phase plan, while
+   *  non-research plugin returns a list of phases to be inserted.
+   */
+  def research: Boolean = false
+
+
+  /** Non-research plugins should override this method to return the phases
+   *
+   *  @param options: commandline options to the plugin, `-P:plugname:opt1,opt2` becomes List(opt1, opt2)
+   *  @return a list of phases to be added to the phase plan
+   */
+  def init(options: List[String]): List[PluginPhase] = ???
+
+  /** Research plugins should override this method to return the new phase plan
+   *
+   *  @param options: commandline options to the plugin, `-P:plugname:opt1,opt2` becomes List(opt1, opt2)
+   *  @param plan: the given phase plan
+   *  @return the new phase plan
+   */
+  def init(options: List[String], plan: List[List[Phase]])(implicit ctx: Context): List[List[Phase]] = ???
+
+  /** A description of this plugin's options, suitable as a response
+   *  to the -help command-line option.  Conventionally, the options
+   *  should be listed with the `-P:plugname:` part included.
+   */
+  val optionsHelp: Option[String] = None
+}
+
+/** ...
+ *
+ *  @author Lex Spoon
+ *  @version 1.0, 2007-5-21
+ */
+object Plugin {
+
+  private val PluginXML = "scalac-plugin.xml"
```

Comment:
IIRC, the "name" in that xml is unused. Maybe it's possible to ditch the xml. See Adriaan's work to avoid dependency on scala-xml at some point. It could be a properties file, or just a file containing a class name, radical simplicity. More like java's service loader. Seems a shame to drag xml config into scala 3.

Reply:
Indeed, only the class name is actually used. Maybe we can assume a text file named `plugins` with each line corresponding to a class name.